### PR TITLE
Extend style guideline documentation

### PR DESCRIPTION
**Pull request by chrisduerr**

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -99,6 +99,13 @@ All Alacritty changes are automatically verified by CI to conform to its rustfmt
 build is failing because of formatting issues, you can install rustfmt using `rustup component add
 rustfmt` and then format all code using `cargo fmt`.
 
+Unless otherwise specified, Alacritty follows the Rust compiler's style guidelines:
+
+https://github.com/rust-lang/api-guidelines
```
Comment (cole-h) [inline CONTRIBUTING.md:104]: I think it would be better to link directly to the book: https://rust-lang.github.io/api-guidelines/

That way, the guidelines are front and foremost (or at least, moreso than before), instead of requiring one to follow another link, or click through the various files in the repo.

File: alacritty/src/config/bindings.rs [hunk 2]
```diff
@@ -452,8 +452,9 @@ pub fn default_key_bindings() -> Vec<KeyBinding> {
             F20,    mods, ~TermMode::VI; Action::Esc(format!("\x1b[34;{}~", modifiers_code));
         ));
 
-        // We're adding the following bindings with `Shift` manually above, so skipping them here
-        // modifiers_code != Shift
+        // We're adding the following bindings with `Shift` manually above, so skipping them here.
+
+        // modifiers_code != Shift.
```
Comment (chrisduerr) [inline alacritty/src/config/bindings.rs:457]: Seems comment is broken-ish?

File: alacritty_terminal/src/ansi.rs [hunk 3]
```diff
@@ -363,65 +363,65 @@ impl Default for CursorStyle {
...
     SgrMouse = 1006,
-    /// ?1007
+    /// ?1007.
     AlternateScroll = 1007,
-    /// ?1049
+    /// ?1049.
     SwapScreenAndSetRestoreCursor = 1049,
-    /// ?2004
+    /// ?2004.
     BracketedPaste = 2004,
```
Comment (chrisduerr) [inline alacritty_terminal/src/ansi.rs:420]: In formatted comments like this it seems counter-productive to add a period at the end.

Wouldn't want people to think it's `?1049.` as escape. If anything, we should do something more like:

At the top there's also a list of things, we shouldn't add periods there.

File: alacritty_terminal/src/ansi.rs [hunk 4]
```diff
@@ -1138,20 +1138,22 @@ where
...
     }
 }
 
 fn attrs_from_sgr_parameters(parameters: &[i64]) -> Vec<Option<Attr>> {
-    // Sometimes a C-style for loop is just what you need
-    let mut i = 0; // C-for initializer
+    // Sometimes a C-style for loop is just what you need.
+
+    // C-for initializer.
+    let mut i = 0;
```
Comment (chrisduerr) [inline alacritty_terminal/src/ansi.rs:1152]: This comment should be reworked. With an extra newline like this it's a bit annoying.

File: alacritty_terminal/src/config/visual_bell.rs [hunk 5]
```diff
@@ -43,15 +43,17 @@ impl VisualBellConfig {
...
+    Ease,
+    EaseOut,
+    // Penner.
+    EaseOutSine,
+    EaseOutQuad,
+    EaseOutCubic,
+    EaseOutQuart,
+    EaseOutQuint,
+    EaseOutExpo,
+    EaseOutCirc,
```
Comment (chrisduerr) [inline alacritty_terminal/src/config/visual_bell.rs:56]: This change doesn't correctly show that the comments affect all elements after the comment.

File: alacritty_terminal/src/grid/tests.rs [hunk 6]
```diff
@@ -311,8 +311,8 @@ fn grow_reflow_multiline() {
     assert_eq!(grid[2][Column(4)], cell('5'));
     assert_eq!(grid[2][Column(5)], cell('6'));
 
-    // Make sure rest of grid is empty
-    // https://github.com/rust-lang/rust-clippy/issues/3788
+    // Make sure rest of grid is empty.
+    // https://github.com/rust-lang/rust-clippy/issues/3788.
```
Comment (chrisduerr) [inline alacritty_terminal/src/grid/tests.rs:315]: I don't think we need to punctuate urls.

File: alacritty_terminal/src/locale.rs [hunk 7]
```diff
@@ -71,9 +71,10 @@ fn system_locale() -> String {
         // `localeIdentifier` returns extra metadata with the locale (including currency and
         // collator) on newer versions of macOS. This is not a valid locale, so we use
         // `languageCode` and `countryCode`, if they're available (macOS 10.12+):
-        // https://developer.apple.com/documentation/foundation/nslocale/1416263-localeidentifier?language=objc
-        // https://developer.apple.com/documentation/foundation/nslocale/1643060-countrycode?language=objc
-        // https://developer.apple.com/documentation/foundation/nslocale/1643026-languagecode?language=objc
+        //
+        // https://developer.apple.com/documentation/foundation/nslocale/1416263-localeidentifier?language=objc.
+        // https://developer.apple.com/documentation/foundation/nslocale/1643060-countrycode?language=objc.
+        // https://developer.apple.com/documentation/foundation/nslocale/1643026-languagecode?language=objc.
```
Comment (chrisduerr) [inline alacritty_terminal/src/locale.rs:77]: Shouldn't punctuate URLs.

File: alacritty_terminal/src/term/color.rs [hunk 8]
```diff
@@ -24,7 +24,7 @@ pub struct Rgb {
     pub b: u8,
 }
 
-// a multiply function for Rgb, as the default dim is just *2/3
+// a multiply function for Rgb, as the default dim is just *2/3.
```
Comment (chrisduerr) [inline alacritty_terminal/src/term/color.rs:27]:
```suggestion
// A multiply function for Rgb, as the default dim is just *2/3.
```

File: alacritty_terminal/src/term/mod.rs [hunk 9]
```diff
@@ -1234,11 +1236,12 @@ impl<T> Term<T> {
     where
         T: EventListener,
     {
-        // Setting 132 column font makes no sense, but run the other side effects
-        // Clear scrolling region
+        // Setting 132 column font makes no sense, but run the other side effects.
+        //
+        // Clear scrolling region.
```
Comment (chrisduerr) [inline alacritty_terminal/src/term/mod.rs:1241]: Why did you change the comment style to this? The newline seems counter-productive for simple comments like this.

File: alacritty_terminal/src/tty/mod.rs [hunk 10]
```diff
@@ -12,7 +12,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 //
-//! tty related functionality
+//! TTY related functionality.
```
Comment (chrisduerr) [inline alacritty_terminal/src/tty/mod.rs:15]: TTY capitalized everywhere, but pty isn't? That seems like a strange choice.

File: alacritty/src/config/bindings.rs (hunk 2 above)
Comment (kchibisov) [reply alacritty/src/config/bindings.rs:457]: I split it intentionally.

Comment (chrisduerr) [reply alacritty/src/config/bindings.rs:457]: I think it could be improved. Without a direct reference to some code, the comment kinda loses its reference point making it a bit unclear what specifically it refers to.

Comment (chrisduerr) [reply alacritty/src/config/bindings.rs:457]: We could also define shift etc as local constants, making the shift documentation redundant.

File: alacritty_terminal/src/ansi.rs [hunk 11]
```diff
@@ -363,29 +363,29 @@ impl Default for CursorStyle {
...
-    /// IRM Insert Mode
+    /// IRM Insert Mode.
     ///
-    /// NB should be part of non-private mode enum
+    /// NB should be part of non-private mode enum.
     ///
-    /// * `CSI 4 h` change to insert mode
-    /// * `CSI 4 l` reset to replacement mode
+    /// * `CSI 4 h` change to insert mode.
+    /// * `CSI 4 l` reset to replacement mode.
```
Comment (chrisduerr) [inline alacritty_terminal/src/ansi.rs:388]: My comments related to this still apply.

File: alacritty_terminal/src/ansi.rs [hunk 12]
```diff
@@ -1138,20 +1138,20 @@ where
...
 fn attrs_from_sgr_parameters(parameters: &[i64]) -> Vec<Option<Attr>> {
-    // Sometimes a C-style for loop is just what you need
-    let mut i = 0; // C-for initializer
+    // C-for initializer.
+    let mut i = 0;
     let mut attrs = Vec::with_capacity(parameters.len());
     loop {
         if i >= parameters.len() {
-            // C-for condition
+            // C-for condition.
```
Comment (chrisduerr) [inline alacritty_terminal/src/ansi.rs:1154]:
```suggestion
```

File: alacritty_terminal/src/ansi.rs [hunk 13]
```diff
@@ -1138,20 +1138,20 @@ where
...
             (b'\\', None) => (),
             _ => unhandled!(),
         }
     }
 }
 
 fn attrs_from_sgr_parameters(parameters: &[i64]) -> Vec<Option<Attr>> {
-    // Sometimes a C-style for loop is just what you need
-    let mut i = 0; // C-for initializer
+    // C-for initializer.
```
Comment (chrisduerr) [inline alacritty_terminal/src/ansi.rs:1149]:
```suggestion
```

Honestly just best to get rid of this probably. Could also rewrite it to a for loop.

File: alacritty_terminal/src/ansi.rs [hunk 14]
```diff
@@ -1231,20 +1231,21 @@ fn attrs_from_sgr_parameters(parameters: &[i64]) -> Vec<Option<Attr>> {
 
         attrs.push(attr);
 
-        i += 1; // C-for expr
+        // C-for expr.
```
Comment (chrisduerr) [inline alacritty_terminal/src/ansi.rs:1234]:
```suggestion
```

File: alacritty_terminal/src/config/visual_bell.rs [hunk 15]
```diff
@@ -43,15 +43,17 @@ impl VisualBellConfig {
...
+    EaseOut,
+    // Penner animations.
+    EaseOutSine,
+    EaseOutQuad,
+    EaseOutCubic,
+    EaseOutQuart,
+    EaseOutQuint,
+    EaseOutExpo,
+    EaseOutCirc,
     Linear,
```
Comment (chrisduerr) [inline alacritty_terminal/src/config/visual_bell.rs:57]: My comments about this are still relevant.

File: alacritty/src/event.rs [hunk 16]
```diff
@@ -738,13 +737,13 @@ impl<N: Notify + OnResize> Processor<N> {
...
 
-    // Write the ref test results to the disk
+    // Write the ref test results to the disk.
     pub fn write_ref_test_results<T>(&self, terminal: &Term<T>) {
         if !self.config.debug.ref_test {
             return;
         }
 
-        // dump grid state
+        // dump grid state.
```
Comment (cole-h) [inline alacritty/src/event.rs:746]:
```suggestion
        // Dump grid state.
```

File: alacritty/src/input.rs [hunk 17]
```diff
@@ -536,7 +536,7 @@ impl<'a, T: EventListener, A: ActionContext<T>> Processor<'a, T, A> {
                 MouseButton::Left => 0,
                 MouseButton::Middle => 1,
                 MouseButton::Right => 2,
-                // Can't properly report more than three buttons.
+                // Can't properly report more than three buttons..
```
Comment (cole-h) [inline alacritty/src/input.rs:539]: Looks accidental.

```suggestion
                // Can't properly report more than three buttons.
```

File: alacritty/src/main.rs [hunk 18]
```diff
@@ -185,50 +185,50 @@ fn run(window_event_loop: GlutinEventLoop<Event>, config: Config) -> Result<(),
...
-    // is dropped, it can ensure Conpty is dropped before the conout pipe in the pty drop order.
+    // The fix is to ensure that processor is dropped first. That way, when io_thread (i.e. PTY)
+    // is dropped, it can ensure ConPTY is dropped before the conout pipe in the PTY drop order.
     //
     // FIXME: Change PTY API to enforce the correct drop order with the typesystem.
     drop(processor);
 
-    // Shutdown PTY parser event loop
-    loop_tx.send(Msg::Shutdown).expect("Error sending shutdown to pty event loop");
+    // Shutdown pty parser event loop.
```
Comment (cole-h) [inline alacritty/src/main.rs:224]:
```suggestion
    // Shutdown PTY parser event loop.
```

File: alacritty/src/renderer/mod.rs [hunk 19]
```diff
@@ -102,34 +102,34 @@ impl From<ShaderCreationError> for Error {
...
 ///
 /// Uniforms are prefixed with "u", and vertex attributes are prefixed with "a".
 #[derive(Debug)]
 pub struct TextShaderProgram {
-    // Program id
+    /// Program id.
     id: GLuint,
 
-    /// projection scale and offset uniform
+    /// projection scale and offset uniform.
```
Comment (cole-h) [inline alacritty/src/renderer/mod.rs:113]:
```suggestion
    /// Projection scale and offset uniform.
```

File: alacritty/src/renderer/mod.rs [hunk 20]
```diff
@@ -225,20 +226,20 @@ impl GlyphCache {
...
 
-        // Load regular font
+        // Load regular font.
         let regular_desc =
             Self::make_desc(&font.normal(), font::Slant::Normal, font::Weight::Normal);
 
         let regular = rasterizer.load_font(&regular_desc, size)?;
 
-        // helper to load a description if it is not the regular_desc
+        // helper to load a description if it is not the regular_desc.
```
Comment (cole-h) [inline alacritty/src/renderer/mod.rs:242]:
```suggestion
        // Helper to load a description if it is not the `regular_desc`.
```

File: alacritty/src/renderer/mod.rs [hunk 21]
```diff
@@ -1276,7 +1277,7 @@ impl RectShaderProgram {
             gl::UseProgram(program);
         }
 
-        // get uniform locations
+        // get uniform locations.
```
Comment (cole-h) [inline alacritty/src/renderer/mod.rs:1280]:
```suggestion
        // Get uniform locations.
```

File: alacritty/src/renderer/mod.rs [hunk 22]
```diff
@@ -1368,60 +1369,60 @@ fn create_shader(
...
     }
 
-    // Build a string
+    // Build a string.
     unsafe {
         buf.set_len(actual_length as usize);
     }
 
-    // XXX should we expect opengl to return garbage?
+    // XXX should we expect OpenGL to return garbage?.
```
Comment (cole-h) [inline alacritty/src/renderer/mod.rs:1413]:
```suggestion
    // XXX should we expect OpenGL to return garbage?
```

File: alacritty/src/renderer/mod.rs [hunk 23]
```diff
@@ -1472,36 +1473,36 @@ impl From<io::Error> for ShaderCreationError {
...
-    /// Width of atlas
+    /// Width of atlas.
     width: i32,
 
-    /// Height of atlas
+    /// Height of atlas.
     height: i32,
 
-    /// Left-most free pixel in a row.
+    /// Left-most free pixel in a row..
```
Comment (cole-h) [inline alacritty/src/renderer/mod.rs:1485]:
```suggestion
    /// Left-most free pixel in a row.
```

File: alacritty/src/renderer/mod.rs [hunk 24]
```diff
@@ -1551,21 +1552,21 @@ impl Atlas {
...
         }
 
-        // If there's still not room, there's nothing that can be done here.
+        // If there's still not room, there's nothing that can be done here..
         if !self.room_in_row(glyph) {
             return Err(AtlasInsertError::Full);
         }
 
-        // There appears to be room; load the glyph.
+        // There appears to be room; load the glyph..
```
Comment (cole-h) [inline alacritty/src/renderer/mod.rs:1565]:
```suggestion
        // There appears to be room; load the glyph.
```

File: alacritty_terminal/src/ansi.rs [hunk 25]
```diff
@@ -142,185 +142,185 @@ impl Processor {
...
-    /// DECKPAM - Set keypad to applications mode (ESCape instead of digits)
+    /// DECKPAM - Set keypad to applications mode (ESCape instead of digits).
     fn set_keypad_application_mode(&mut self) {}
 
-    /// DECKPNM - Set keypad to numeric mode (digits instead of ESCape seq)
+    /// DECKPNM - Set keypad to numeric mode (digits instead of ESCape seq).
     fn unset_keypad_application_mode(&mut self) {}
 
-    /// Set one of the graphic character sets, G0 to G3, as the active charset.
+    /// Set one of the graphic character sets, G0 to G3, as the active charset..
```
Comment (cole-h) [inline alacritty_terminal/src/ansi.rs:302]:
```suggestion
    /// Set one of the graphic character sets, G0 to G3, as the active charset.
```

File: alacritty_terminal/src/grid/tests.rs [hunk 26]
```diff
@@ -156,7 +156,7 @@ fn test_iter() {
     assert_eq!(Some(&3), iter.next());
     assert_eq!(Some(&4), iter.next());
 
-    // test linewrapping
+    // test line-wrapping.
```
Comment (cole-h) [inline alacritty_terminal/src/grid/tests.rs:159]:
```suggestion
    // Test line-wrapping.
```

File: alacritty_terminal/src/grid/tests.rs [hunk 27]
```diff
@@ -165,10 +165,10 @@ fn test_iter() {
     assert_eq!(Column(4), iter.point().col);
     assert_eq!(4, iter.point().line);
 
-    // Make sure iter.cell() returns the current iterator position
+    // Make sure iter.cell() returns the current iterator position.
     assert_eq!(&4, iter.cell());
 
-    // test that iter ends at end of grid
+    // test that iter ends at end of grid.
```
Comment (cole-h) [inline alacritty_terminal/src/grid/tests.rs:171]:
```suggestion
    // Test that iter ends at end of grid.
```

File: font/src/darwin/mod.rs [hunk 28]
```diff
@@ -156,21 +156,21 @@ impl crate::Rasterize for Rasterizer {
...
 
-        // first try the font itself as a direct hit
+        // First try the font itself as a direct hit.
         self.maybe_get_glyph(glyph, font).unwrap_or_else(|| {
-            // then try fallbacks
+            // Then try fallbacks.
             for fallback in &font.fallbacks {
                 if let Some(result) = self.maybe_get_glyph(glyph, &fallback) {
-                    // found a fallback
+                    // found a fallback.
```
Comment (cole-h) [inline font/src/darwin/mod.rs:169]:
```suggestion
                    // Found a fallback.
```
